File: source/py/game_of_life.py
import random
import time
import copy


HEIGHT = 10
WIDTH = 10
FPS_LIMIT = 30


class GameOfLife:
    def __init__(self, height=HEIGHT, width=HEIGHT):
        self.height = height
        self.width = width
        self.grid = []
        self.init_grid()

    def init_grid(self):
        # Willekeurige initialisatie
        self.grid = [
            [random.choice([True, False]) for _ in range(self.width)]
            for _ in range(self.height)
        ]

        # Randen altijd dood
        for i in range(self.height):
            self.grid[i][0] = False
            self.grid[i][self.width - 1] = False

        for j in range(self.width):
            self.grid[0][j] = False
            self.grid[self.height - 1][j] = False

    def run(self):
        previous_state = None
        frame_duration = 1 / FPS_LIMIT
        frame = 0

        while True:
            self.print_grid(self.grid)

            frame += 1
            print(f"Frame: {frame}")

            next_state = self.get_next_state(self.grid)

            if self.check_stable(previous_state, self.grid, next_state):
                print("System is stable. Stopping.")
                break

            previous_state = copy.deepcopy(self.grid)
            self.grid = next_state

            time.sleep(frame_duration)

    def get_next_state(self, current_state):
        # Nieuwe lege grid
        new_state = [
            [False for _ in range(self.width)]
            for _ in range(self.height)
        ]

        for i in range(1, self.height - 1):
            for j in range(1, self.width - 1):
                neighbours = self.count_neighbours(current_state, i, j)

                if current_state[i][j]:
                    # Levend
                    if neighbours == 2 or neighbours == 3:
                        new_state[i][j] = True
                else:
                    # Dood
                    if neighbours == 3:
                        new_state[i][j] = True

        return new_state

    def count_neighbours(self, state, i, j):
        counter = 0

        for y in range(-1, 2):
            for x in range(-1, 2):
                if y == 0 and x == 0:
                    continue

                if state[i + y][j + x]:
                    counter += 1

        return counter

    def check_stable(self, previous_state, current_state, next_state):
        if previous_state is None:
            return False

        for i in range(self.height):
            for j in range(self.width):
                if (previous_state[i][j] != current_state[i][j] or
                        previous_state[i][j] != next_state[i][j]):
                    return False
        return True

    def print_grid(self, grid):
        # Console leegmaken (werkt op meeste systemen)
        print("\033[H\033[2J", end="")

        for row in grid:
            for cell in row:
                print("#" if cell else ".", end="")
            print()
        print()


if __name__ == "__main__":
    print("Starting Game of Life...")
    game = GameOfLife()
    game.run()

File: source/js/game_of_life.js
const HEIGHT = 10;
const WIDTH = 10;
const FPS_LIMIT = 30;

class GameOfLife {
  constructor(height = HEIGHT, width = WIDTH) {
    this.height = height;
    this.width = width;
    this.grid = [];
    this.initGrid();
  }

  initGrid() {
    // Willekeurige initialisatie
    this.grid = Array.from({ length: this.height }, () =>
      Array.from({ length: this.width }, () => Math.random() < 0.5),
    );

    // Randen altijd dood
    for (let i = 0; i < this.height; i++) {
      this.grid[i][0] = false;
      this.grid[i][this.width - 1] = false;
    }

    for (let j = 0; j < this.width; j++) {
      this.grid[0][j] = false;
      this.grid[this.height - 1][j] = false;
    }
  }

  async run() {
    let previousState = null;
    const frameDuration = 1000 / FPS_LIMIT;
    let frame = 0;

    while (true) {
      this.printGrid();

      frame++;
      console.log("Frame:", frame);

      const nextState = this.getNextState(this.grid);

      if (this.checkStable(previousState, this.grid, nextState)) {
        console.log("System is stable. Stopping.");
        break;
      }

      previousState = this.deepCopy(this.grid);
      this.grid = nextState;

      await this.sleep(frameDuration);
    }
  }

  getNextState(currentState) {
    const newState = Array.from({ length: this.height }, () =>
      Array.from({ length: this.width }, () => false),
    );

    for (let i = 1; i < this.height - 1; i++) {
      for (let j = 1; j < this.width - 1; j++) {
        const neighbours = this.countNeighbours(currentState, i, j);

        if (currentState[i][j]) {
          if (neighbours === 2 || neighbours === 3) {
            newState[i][j] = true;
          }
        } else {
          if (neighbours === 3) {
            newState[i][j] = true;
          }
        }
      }
    }

    return newState;
  }

  countNeighbours(state, i, j) {
    let counter = 0;

    for (let y = -1; y <= 1; y++) {
      for (let x = -1; x <= 1; x++) {
        if (y === 0 && x === 0) continue;

        if (state[i + y][j + x]) {
          counter++;
        }
      }
    }

    return counter;
  }

  checkStable(previousState, currentState, nextState) {
    if (!previousState) return false;

    for (let i = 0; i < this.height; i++) {
      for (let j = 0; j < this.width; j++) {
        if (
          previousState[i][j] !== currentState[i][j] ||
          previousState[i][j] !== nextState[i][j]
        ) {
          return false;
        }
      }
    }

    return true;
  }

  printGrid() {
    // Console clear (Node.js)
    console.clear();

    for (let i = 0; i < this.height; i++) {
      let row = "";
      for (let j = 0; j < this.width; j++) {
        row += this.grid[i][j] ? "#" : ".";
      }
      console.log(row);
    }
    console.log();
  }

  deepCopy(grid) {
    return grid.map((row) => [...row]);
  }

  sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

(async () => {
  console.log("Starting Game of Life...");
  const game = new GameOfLife();
  await game.run();
})();
